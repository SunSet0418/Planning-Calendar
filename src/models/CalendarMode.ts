export enum CalendarMode {
    Month = '월',
    Week = '주'
}