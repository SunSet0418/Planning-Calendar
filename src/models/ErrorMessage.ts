export class ErrorMessage {
    constructor(
        public title: string,
        public description: string
    ) { }
}