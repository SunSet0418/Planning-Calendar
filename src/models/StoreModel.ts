import { TaskStoreModel } from '../modules/reducers/task';

export interface StoreModel {
    task: TaskStoreModel
}