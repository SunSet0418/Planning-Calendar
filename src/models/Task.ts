export interface Task {
    id: number
    startDate: string,
    endDate: string,
    startTime: number,
    durationTime: number
}