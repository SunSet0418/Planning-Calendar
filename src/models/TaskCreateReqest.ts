export interface TaskCreateRequest {
    name: string,
    date: string,
    startHour: number,
    endHour: number
}