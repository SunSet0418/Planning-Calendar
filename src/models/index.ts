export { TaskStore } from './TaskStore';
export { Task } from './Task';
export { StoreModel } from './StoreModel';
export { CalendarMode } from './CalendarMode';
export { ErrorMessage } from './ErrorMessage';