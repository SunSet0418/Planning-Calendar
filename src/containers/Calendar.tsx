import * as React from 'react';
import * as moment from 'moment';
import styled from 'styled-components';
import { CalendarMode } from '../models/CalendarMode';
import { WeeklyCalendar } from '../components/weekCalendar/WeeklyCalenldar';
import { MonthCalendar } from '../components/monthCalendar/MonthCalendar';
import { Task } from '../models';

const Container = styled.div`
    display: flex;
    flex-flow: column;
    padding: 0 10px;
`;

interface OwnProps {
    currentMoment: moment.Moment,
    mode: CalendarMode,
    tasks: Task[],
    onSelect: (selectedMoment?: moment.Moment | Task) => void,
}

type Props = OwnProps & React.HTMLAttributes<HTMLDivElement>;

export const Calendar: React.FC<Props> = ({currentMoment, mode, tasks, onSelect, ...divProps}) => {

    const handleClick = (e: React.MouseEvent<HTMLDivElement, MouseEvent>) => {
        const target = e.target as HTMLElement;
        if(target.dataset.datetime) {
            onSelect(moment(target.dataset.datetime))
        } else if(target.dataset.taskid) {
            onSelect(tasks.find(task => task.id === Number(target.dataset.taskid)));
        }
    }

    return (
        <Container {...divProps} onClick={handleClick}>
            {mode === CalendarMode.Month && (
                <MonthCalendar
                    style={{ flex: 1 }}
                    currentMoment={currentMoment}
                    tasks={tasks}
                />
            )}
            {mode === CalendarMode.Week && (
                <WeeklyCalendar
                    style={{ flex: 1 }}
                    currentMoment={currentMoment}
                    tasks={tasks}
                />
            )}
        </Container>
    )
}
