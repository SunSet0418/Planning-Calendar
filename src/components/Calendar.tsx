import * as React from 'react';
import * as moment from 'moment';
import { useState } from 'react'; 
import styled from 'styled-components';
import { CalendarMode } from '../models/CalendarMode';
import { Task, TaskCreateRequest } from '../models/Task';
import update from 'immutability-helper';
import { WeeklyCalendar } from './weekCalendar/WeeklyCalenldar';
import { MonthCalendar } from './monthCalendar/MonthCalendar';
import { TaskModal } from './TaskModal';

const Container = styled.div`
    display: flex;
    flex-flow: column;
    padding: 0 10px;
`;

interface OwnProps {
    currentMoment: moment.Moment,
    mode: CalendarMode
}

type Props = OwnProps & React.HTMLAttributes<HTMLDivElement>;

export const Calendar: React.FC<Props> = (props: Props) => {
    const { currentMoment, mode, ...divProps } = props;
    const [tasks, setTasks] = useState(new Array<Task>());
    const [selectedMoment, setSelectedMoment] = useState();
    const [selectedTask, setSelectedTask] = useState();

    const handleClick = (e: React.MouseEvent<HTMLDivElement, MouseEvent>) => {
        const datetime = (e.target as HTMLElement).dataset.datetime;
        const taskId = (e.target as HTMLElement).dataset.taskid;
        if (datetime) {
            setSelectedMoment(moment(datetime));
        } else if (taskId) {
            setSelectedTask(tasks.find(task => task.id === Number(taskId)));
        }
    }
    const handleDismissModal = () => {
        setSelectedMoment(null);
        setSelectedTask(null);
    }
    const handleUpdateTask = (task: TaskCreateRequest) => {
        if (selectedTask) {
            const idx = tasks.findIndex(t => t.id === selectedTask.id);
            setTasks(update(tasks, { $splice: [[idx, 1, { id: selectedTask.id, ...task }]] }));
        } else {
            task.id = tasks.length;
            setTasks(update(tasks, {$push: [task]}));
        }
        handleDismissModal();
    }
    const handleDeleteTask = () => {
        const idx = tasks.findIndex(t => t.id === selectedTask.id);
        setTasks(update(tasks, {$splice: [[idx, 1]]}));
        handleDismissModal();
    }
    return (
        <Container {...divProps} onClick={handleClick}>
            {mode === CalendarMode.Month && (
                <MonthCalendar
                    style={{ flex: 1 }}
                    currentMoment={props.currentMoment}
                    tasks={tasks}
                />
            )}
            {mode === CalendarMode.Week && (
                <WeeklyCalendar
                    style={{ flex: 1 }}
                    currentMoment={props.currentMoment}
                    tasks={tasks}
                />
            )}
            {(selectedTask || selectedMoment) && (
                <TaskModal
                    task={selectedTask}
                    time={selectedMoment}
                    onOk={handleUpdateTask}
                    onCancel={handleDismissModal}
                    onDelete={handleDeleteTask} />
            )}
        </Container>
    )
}