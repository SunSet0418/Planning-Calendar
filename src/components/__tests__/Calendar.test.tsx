import * as React from 'react';
import * as Enzyme from 'enzyme';
import { Calendar } from '../Calendar';
import { CalendarMode } from '../../models/CalendarMode';
import * as moment from 'moment';

describe('<Calendar />', () => {
    const testMoment = moment.parseZone('2014-02-27T00:00:00Z');
    let component = null;
    const tasks = [];
    const mockFn = jest.fn();

    it('renders correctly', () => {
        component = Enzyme.shallow(<Calendar currentMoment={testMoment} mode={CalendarMode.Month} tasks={tasks} onSelect={mockFn} onUpdate={mockFn}/>);
    });

    it('should match snapshot', () => {
        expect(component).toMatchSnapshot();
    });
});