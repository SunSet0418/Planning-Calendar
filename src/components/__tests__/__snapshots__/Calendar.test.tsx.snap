// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Calendar /> should match snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Unknown
    currentMoment={"2014-02-27T00:00:00.000Z"}
    mode="ì›”"
    onSelect={[Function]}
    onUpdate={[Function]}
    tasks={Array []}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <Unknown
          currentMoment={"2014-02-27T00:00:00.000Z"}
          style={
                    Object {
                              "flex": 1,
                            }
          }
          tasks={Array []}
/>,
        false,
      ],
      "onClick": [Function],
      "onDragOver": [Function],
      "onDragStart": [Function],
      "onDrop": [Function],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "currentMoment": "2014-02-27T00:00:00.000Z",
          "style": Object {
            "flex": 1,
          },
          "tasks": Array [],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      false,
    ],
    "type": Object {
      "$$typeof": Symbol(react.forward_ref),
      "attrs": Array [],
      "componentStyle": ComponentStyle {
        "componentId": "sc-VigVT",
        "isStatic": true,
        "rules": Array [
          "
    display: flex;
    flex-flow: column;
    padding: 0 10px;
",
        ],
      },
      "displayName": "styled.div",
      "foldedComponentIds": Array [],
      "render": [Function],
      "styledComponentId": "sc-VigVT",
      "target": "div",
      "toString": [Function],
      "warnTooManyClasses": [Function],
      "withComponent": [Function],
    },
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <Unknown
            currentMoment={"2014-02-27T00:00:00.000Z"}
            style={
                        Object {
                                    "flex": 1,
                                  }
            }
            tasks={Array []}
/>,
          false,
        ],
        "onClick": [Function],
        "onDragOver": [Function],
        "onDragStart": [Function],
        "onDrop": [Function],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "currentMoment": "2014-02-27T00:00:00.000Z",
            "style": Object {
              "flex": 1,
            },
            "tasks": Array [],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        false,
      ],
      "type": Object {
        "$$typeof": Symbol(react.forward_ref),
        "attrs": Array [],
        "componentStyle": ComponentStyle {
          "componentId": "sc-VigVT",
          "isStatic": true,
          "rules": Array [
            "
    display: flex;
    flex-flow: column;
    padding: 0 10px;
",
          ],
        },
        "displayName": "styled.div",
        "foldedComponentIds": Array [],
        "render": [Function],
        "styledComponentId": "sc-VigVT",
        "target": "div",
        "toString": [Function],
        "warnTooManyClasses": [Function],
        "withComponent": [Function],
      },
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
